src/Avalonia.Controls/ThemeVariantWindowIcon.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.IO;
+using Avalonia.Platform;
+
+namespace Avalonia.Controls;
+
+/// <summary>
+/// Represents two icons, one for use with a light theme and the other for use with a dark theme.
+/// </summary>
+/// <remarks>
+/// A window's icon can appear in multiple locations (e.g. in the window title bar and the platform taskbar), and these locations 
+/// can use different themes. This means that both light and dark icons can be visible at the same time.
+/// </remarks>
+public class ThemeVariantWindowIcon : WindowIcon
+{
+    private static IWindowIconImpl Extract(WindowIcon icon, string paramName) => icon switch
+    {
+        ThemeVariantWindowIcon => throw new ArgumentException($"{nameof(ThemeVariantWindowIcon)} may not recursively contain further {nameof(ThemeVariantWindowIcon)}.", paramName),
+        null => throw new ArgumentNullException(paramName),
+        _ => icon.PlatformImpl,
+    };
+
+    /// <inheritdoc cref="ThemeVariantWindowIcon"/>
+    public ThemeVariantWindowIcon(WindowIcon light, WindowIcon dark) : base(new Impl(Extract(light, nameof(light)), Extract(dark, nameof(dark))))
+    {
+    }
+
+    private class Impl : IThemeVariantWindowIconImpl
+    {
+        public Impl(IWindowIconImpl light, IWindowIconImpl dark)
+        {
+            Light = light;
+            Dark = dark;
+        }
+
+        public IWindowIconImpl Light { get; }
+        public IWindowIconImpl Dark { get; }
+
+        public void Save(Stream outputStream) => throw new NotSupportedException();
+
+    }
+}
+
+public class ThemeVariantWindowIconExtension
```

Comment:
This `ThemeVariantWindowIcon` and `ThemeVariantWindowIconExtension` API raises confusion.
Why do we need this limited API, when we can define icons in theme dictionaries, just like any other theme specific resource?
```xaml
<ResourceDictionary.ThemeDictionaries>
    <ResourceDictionary x:Key="Light"> <!-- or Default -->
         <WindowIcon x:Key="MyIcon">avares://Project/Assets/LightIcon.ico</WindowIcon>
    </ResourceDictionary>
    <ResourceDictionary x:Key="Dark">
         <WindowIcon x:Key="MyIcon">avares://Project/Assets/DarkIcon.ico</WindowIcon>
    </ResourceDictionary>
</ResourceDictionary.ThemeDictionaries>
```

And then use this icon like `Icon="{DynamicResource MyIcon}"`

Reply:
Note: I don't know for sure if it's as easy to create WindowIcon this way in XAML. If not - it should be easy to implement.

Reply:
I guess this is the same issue as [the previous comment](https://github.com/AvaloniaUI/Avalonia/pull/12107#pullrequestreview-1756737948). The type needs a more distinctive name.